charts/opentelemetry-operator/templates/service.yaml

```diff
@@ -19,16 +19,17 @@ spec:
       protocol: TCP
       targetPort: metrics
   selector:
-    app.kubernetes.io/name: opentelemetry-operator
-    control-plane: controller-manager
+      {{- include "opentelemetry-operator.selectorLabels" . | nindent 6 }}
+      app.kubernetes.io/component: controller-manager
```

Comment:
@ngnix can you test what happens if a user tries to do `helm upgrade` with this version of the chart?  I suspect the changed selectors will cause it to fail.

Reply:
Yes, are I right.

Label Selectors is immutable field and cannot be changed.

```
17:30 $ helm upgrade --install  opentelemetry-operator  . -n opentelemetry --wait --create-namespace 
Error: UPGRADE FAILED: cannot patch "opentelemetry-operator-controller-manager" with kind Deployment: Deployment.apps "opentelemetry-operator-controller-manager" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{"app.kubernetes.io/component":"controller-manager", "app.kubernetes.io/instance":"opentelemetry-operator", "app.kubernetes.io/name":"opentelemetry-operator"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable
```

I do not consider this a problem, because the application itself is completely stateless, removing the operator does not remove the CRD, therefore reinstalling the operator will not cause serious problems.

Perhaps this MR should immediately add support for fullnameOverride, since this change will also affect immutable fields

Reply:
```
17:40 $ kubectl -n opentelemetry delete deployments.apps -l app.kubernetes.io/name=opentelemetry-operator --cascade=orphan
deployment.apps "opentelemetry-operator-controller-manager" deleted
17:40 $ kubectl -n opentelemetry get po 
NAME                                                         READY   STATUS    RESTARTS   AGE
opentelemetry-operator-controller-manager-7858d54d75-jgshd   2/2     Running   0          11m

$ helm upgrade --install  opentelemetry-operator  . -n opentelemetry --wait --create-namespace 
Release "opentelemetry-operator" has been upgraded. Happy Helming!
NAME: opentelemetry-operator
LAST DEPLOYED: Fri Nov 18 17:41:28 2022
NAMESPACE: opentelemetry
STATUS: deployed
REVISION: 4
NOTES:
opentelemetry-operator has been installed. Check its status by running:
  kubectl --namespace opentelemetry get pods -l "release=opentelemetry-operator"

Visit https://github.com/open-telemetry/opentelemetry-operator for instructions on how to create & configure OpenTelemetryCollector and Instrumentation custom resources by using the Operator.

[Important] The v0.17.0 helm chart version changes OpenTelemetry Collector image to the contrib version. If you want to use the core version, set `manager.collectorImage.repository` to `otel/opentelemetry-collector`.


$ kubectl -n opentelemetry get po 
NAME                                                         READY   STATUS    RESTARTS   AGE
opentelemetry-operator-controller-manager-7858d54d75-jgshd   2/2     Running   0          11m
opentelemetry-operator-controller-manager-86c74dc4f9-fpcjg   2/2     Running   0          14s

$ kubectl -n opentelemetry get replicasets.apps 
NAME                                                   DESIRED   CURRENT   READY   AGE
opentelemetry-operator-controller-manager-7858d54d75   1         1         1       12m
opentelemetry-operator-controller-manager-86c74dc4f9   1         1         1       43s


$ kubectl -n opentelemetry delete replicasets.apps opentelemetry-operator-controller-manager-7858d54d75 
replicaset.apps "opentelemetry-operator-controller-manager-7858d54d75" deleted

$ kubectl -n opentelemetry get po 
NAME                                                         READY   STATUS    RESTARTS   AGE
opentelemetry-operator-controller-manager-86c74dc4f9-fpcjg   2/2     Running   0          58s
```

As a solution, you can attach an old and dirty hack to the documentation :

1. manual delete deployment with option `--cascade=orphan`, this command remove deployment but not remove replicaset and pods
2. install new chart version, this step create new deployment, replicaset and pods
3. manual delete older replicaset

Once upon a time, I personally did this with [Prometheus](https://github.com/prometheus-community/helm-charts/blob/main/charts/prometheus/README.md#to-150)

Reply:
@ngnix when introducing a breaking change you should bump the chart to a new major version (0.19) and document the steps that the user needs to do in https://github.com/open-telemetry/opentelemetry-helm-charts/blob/main/charts/opentelemetry-operator/UPGRADING.md

Reply:
Ok,  I bump new version after add support for fullnameOverride,  to avoid breaking changes in the future.

Reply:
@ngnix you can add fullnameoverride support now or it can be done later.  Either way we can handle the breaking change with the chart bump and the UPGRADING doc like @Allex1 mentioned.

Reply:
@TylerHelmuth I'll add support and test locally tomorrow afternoon + I'll upgrade to 19.0.0

Reply:
@ngnix please bump to `0.19.0`.  @Allex1 meant to say `minor version` not `major version`.

Reply:
Hi @Allex1 @TylerHelmuth , I was bump chart version and generate new examples, now CI is red, I didn't really understand from the CI logs what the problem is, can you explain?

Reply:
@ngnix ci now passing. Pls resolve conflict via a rebase and `make generate-examples`  so we can merge

Reply:
@Allex1 conflicts resolved